app/jobs/reports/monthly_account_reuse_report.rb

```diff
@@ -0,0 +1,133 @@
+        end
+      end
+  
+      def first_day_of_report_month
+        report_date.strftime("%Y-%m-01")
+      end
+
+      def stats_month
+        report_date.prev_month(1).strftime("%b-%Y")
+      end
+  
+      def report_body
+        params = {
+          query_date: first_day_of_report_month,
+          report_month: stats_month
+        }.transform_values { |v| ActiveRecord::Base.connection.quote(v) }
+  
+        agency_sql = format(<<-SQL, params)
+          SELECT
+              COUNT(*) AS num_users
+          , agencies_per_user.num_agencies
+          FROM (
+              SELECT
+                  COUNT(DISTINCT agencies.id) AS num_agencies
+              , identities.user_id
+              FROM 
+                  identities
+              JOIN 
+                  service_providers sp ON identities.service_provider = sp.issuer
+              JOIN 
+                  agencies ON sp.agency_id = agencies.id
+              WHERE
+                  identities.last_ial2_authenticated_at IS NOT NULL
+              AND
+                  identities.verified_at < %{query_date}
+              GROUP BY 
+                  identities.user_id
+          ) agencies_per_user
+          GROUP BY 
+              agencies_per_user.num_agencies
+          HAVING agencies_per_user.num_agencies > 1
+          ORDER BY 
+              num_agencies ASC
+        SQL
+
+        proofed_sql = format(<<-SQL, params)
+          SELECT
+            COUNT(*) AS num_proofed
+          FROM
+            profiles
+          WHERE
+            profiles.active = TRUE
+          AND
+            profiles.activated_at < %{query_date}
+        SQL
+  
+        agency_results = transaction_with_timeout do
+          ActiveRecord::Base.connection.execute(agency_sql)
+        end
+
+        proofed_results = transaction_with_timeout do
+          ActiveRecord::Base.connection.execute(proofed_sql)
+        end
+        
+        reuse_report = agency_results.as_json
+
+        total_proofed = proofed_results[0]['num_proofed']
+
+        total_reuse_stats = {
+          label: "Total (all >1)",
+          num_users: 0,
+          percentage: 0
+        }
+
+        reuse_report.each do |result_entry|
+          total_reuse_stats.num_users += result_entry.num_users
+        end
+
+        reuse_report.each_with_index do |result_entry, index|
+          reuse_report[index].percentage = result_entry.num_users/total_proofed*100
+          total_reuse_stats.percentage += reuse_report[index].percentage
+        end
+
+        report_csv = CSV.generate do |csv|
+            csv << [
+            "#{stats_month} Num. SPs",
+            "#{stats_month} Num. users",
+            "#{stats_month} Reuse Rate Percentage",
+            ] 
+    
+            reuse_report.each do |result_entry|
+            csv << [
+              result_entry.num_agencies,
+              result_entry.num_users,
+              result_entry.percentage,
+            ]
+            end
+        end
```

Comment:
I would start breaking these out into separate methods

Reply:
Continued in another thread